containers/src/Data/Set/Internal.hs

```diff
@@ -594,6 +599,74 @@ delete = go
 {-# INLINE delete #-}
 #endif
 
+-- | /O(log n)/ @('alterF' f k s)@ can delete or insert the key @k@ in @s@
+-- depending on its pre-existing presence in @s@.
+--
+-- In short: @'member' k \<$\> 'alterF' f k s = f ('member' k s)@.
+alterF :: (Ord a, Functor f) => (Bool -> f Bool) -> a -> Set a -> f (Set a)
+alterF f k s = fmap choose (f member_)
+  where
+    (member_, inserted, deleted) = case alteredSet k s of
+        Deleted d           -> (True , s, d)
+        Inserted i          -> (False, i, s)
+        DeletedInserted d i -> (True , i, d)
+
+    choose True  = inserted
+    choose False = deleted
+#ifndef __GLASGOW_HASKELL__
+{-# INLINE alterF #-}
+#else
+{-# INLINABLE [2] alterF #-}
+
+{-# RULES
+"alterF/Const" forall k (f :: Bool -> Const a Bool) . alterF f k = \s -> Const . getConst . f $ member k s
+ #-}
```

Comment:
I wonder whether I should try to add rules for `insert` (`const (Identity True)`) and `delete` (`const (Identity False)`).

I'm not very experienced with rules but I've seen that `bytestring` has rules scrutinizing function arguments: https://github.com/haskell/bytestring/blob/95fe6bdf13c9cc86c1c880164f7844d61d989574/Data/ByteString.hs#L859-L864

Reply:
I'd start with the more general rule:

```haskell
alterF (const m) k s = (bool (delete k s) (insert k s)) <$> m
```

You'll have to be clever about the rules. The basic trick is to use rules that look something roughly like this:

```haskell
alterF' f _ _ k s = alterF f k s
-- Early rules
alterF f k s = alterF' f (f False) (f True) k s
alterF' f m m k s =  (bool (delete k s) (insert k s)) <$> m
-- Late rule
alterF' f m n k s = alterF f k s
```

Much poring over Core will be necessary! I haven't included the necessary inline staging.

Reply:
(Last comment has been edited.)

Reply:
Hm, I'm honestly not very motivated to put much more work into this. Performance for `insert` and `delete` looks already reasonable, and that's with a custom `Ident` `Functor` where the `SPECIALIZE` below doesn't apply:

```
benchmarked insert
time                 1.090 ms   (1.076 ms .. 1.101 ms)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 1.092 ms   (1.086 ms .. 1.102 ms)
std dev              25.13 μs   (16.94 μs .. 37.69 μs)

benchmarked alterF:insert
time                 1.261 ms   (1.245 ms .. 1.274 ms)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 1.255 ms   (1.248 ms .. 1.263 ms)
std dev              25.63 μs   (20.42 μs .. 31.36 μs)
```

```
benchmarked delete
time                 493.4 μs   (491.7 μs .. 495.8 μs)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 499.4 μs   (497.6 μs .. 503.8 μs)
std dev              8.814 μs   (4.853 μs .. 16.90 μs)

benchmarked alterF:delete
time                 570.6 μs   (568.5 μs .. 573.8 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 570.3 μs   (569.9 μs .. 570.9 μs)
std dev              1.509 μs   (979.5 ns .. 2.688 μs)
```

Also, in order to preserve keys instead of replacing them, we'd need a custom `insert` variant.

Reply:
I can make an issue so someone more motivated can work on those rules, if you want, @treeowl.